### ISSUE TITLE
Encrtpted-CNN training

### ISSUE THREAD

**Issue by Samuel-1029** [OpenMined/TenSEAL#382]
## Question
**Describe your question in ONE SENTENCE.**
How can i use the loss to implement the backward of PyTorch in the process of CNN training?
## Further Information
Describe your question in greater length here.

## Screenshots
If applicable, add screenshots to help explain your question.

## System Information
 - OS Version: [e.g. 22]
 - Language Version: [e.g. Python 3.7, Node 10.18.1]
 - Package Manager Version: [e.g. Conda 4.6.1, NPM 6.14.1]

**Comment by basavyr** [comment]
@Samuel-1029  it depends on your task, and therefore on the Loss function you want to use in your training.
As far as I know, the current state of the library does not support encrypted operations for loss-functions that are specific to classification tasks, such as Cross-Entropy Loss, Negative-Log-Likelihood loss, and so on. As such, you would have to code the functions yourself. Moreover, keep in mind that operations like `Softmax`  (required for losses) do not work out-of-the-box, which means that you either perform `Softmax` on decrypted data or find a way to approximate it through a polynomial.

It really depends on your expression that you consider for the cost function. I myself struggle with the implementation of the [CELoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) to work with encrypted data.

I can show you how the method looks like:
```python
def ce_loss(self, y_pred_enc: ts.CKKSTensor, y_true_onehot: torch.Tensor) -> ts.CKKSTensor:
        y_pred_dec = y_pred_enc.decrypt()
        # we need to decrypt the predicted output (which is defined as a CKKS tensor)
        y_pred_torch = torch.tensor(y_pred_dec.tolist()).view(y_pred_enc.shape)
        log_softmax = F.log_softmax(y_pred_torch, dim=1)

        loss = -torch.sum(log_softmax*y_true_onehot,
                          dim=1).mean().unsqueeze(-1)

        loss_enc = ts.ckks_tensor(self.ctx, loss)
        return loss_enc
``` 

Here I am decrypting the `CKKS` tensor and apply the PyTorch implementation and finally encrypting back the final loss value.

I hope this helps you.
